csharp/ql/src/semmle/code/csharp/dataflow/SignAnalysis.qll

```diff
@@ -0,0 +1,595 @@
+}
+
+/**
+ * Holds if `bound` is a bound for `v` at `pos` that needs to be negative in
+ * order for `v` to be negative.
+ */
+private predicate negBound(Expr bound, Definition v, SsaReadPosition pos, boolean isStrict) {
+  upperBound(bound, v, pos, isStrict)
+  or
+  eqBound(bound, v, pos, true) and isStrict = false
+}
+
+/**
+ * Holds if `bound` is a bound for `v` at `pos` that can restrict whether `v`
+ * can be zero.
+ */
+private predicate zeroBound(Expr bound, Definition v, SsaReadPosition pos) {
+  lowerBound(bound, v, pos, _) or
+  upperBound(bound, v, pos, _) or
+  eqBound(bound, v, pos, _)
+}
+
+/** Holds if `bound` allows `v` to be positive at `pos`. */
+private predicate posBoundOk(Expr bound, Definition v, SsaReadPosition pos) {
+  posBound(bound, v, pos, _) and TPos() = exprSign(bound)
+  or
+  posBound(bound, v, pos, true) and TZero() = exprSign(bound)
+}
+
+/** Holds if `bound` allows `v` to be negative at `pos`. */
+private predicate negBoundOk(Expr bound, Definition v, SsaReadPosition pos) {
+  negBound(bound, v, pos, _) and TNeg() = exprSign(bound)
+  or
+  negBound(bound, v, pos, true) and TZero() = exprSign(bound)
+}
+
+/** Holds if `bound` allows `v` to be zero at `pos`. */
+private predicate zeroBoundOk(Expr bound, Definition v, SsaReadPosition pos) {
+  lowerBound(bound, v, pos, _) and TNeg() = exprSign(bound)
+  or
+  lowerBound(bound, v, pos, false) and TZero() = exprSign(bound)
+  or
+  upperBound(bound, v, pos, _) and TPos() = exprSign(bound)
+  or
+  upperBound(bound, v, pos, false) and TZero() = exprSign(bound)
+  or
+  eqBound(bound, v, pos, true) and TZero() = exprSign(bound)
+  or
+  eqBound(bound, v, pos, false) and TZero() != exprSign(bound)
+}
+
+/**
+ * Holds if there is a bound that might restrict whether `v` has the sign `s`
+ * at `pos`.
+ */
+predicate hasGuard(Definition v, SsaReadPosition pos, Sign s) {
+  s = TPos() and posBound(_, v, pos, _)
+  or
+  s = TNeg() and negBound(_, v, pos, _)
+  or
+  s = TZero() and zeroBound(_, v, pos)
+}
+
+// pragma[noinline]
+// private Sign guardedSsaSign(Definition v, SsaReadPosition pos) {
+//   //result = ssaDefSign(v) and
+//   pos.hasReadOfVar(v) and
+//   hasGuard(v, pos, result)
+// }
+pragma[noinline]
+private Sign unguardedSsaSign(Definition v, SsaReadPosition pos) {
+  result = ssaDefSign(v) and
+  pos.hasReadOfVar(v) and
+  not hasGuard(v, pos, result)
+}
+
+private Sign guardedSsaSignOk(Definition v, SsaReadPosition pos) {
+  result = TPos() and
+  forex(Expr bound | posBound(bound, v, pos, _) | posBoundOk(bound, v, pos))
+  or
+  result = TNeg() and
+  forex(Expr bound | negBound(bound, v, pos, _) | negBoundOk(bound, v, pos))
+  or
+  result = TZero() and
+  forex(Expr bound | zeroBound(bound, v, pos) | zeroBoundOk(bound, v, pos))
+}
+
+/** Gets a possible sign for `v` at `pos`. */
+Sign ssaSign(Definition v, SsaReadPosition pos) {
+  result = unguardedSsaSign(v, pos)
+  or
+  //result = guardedSsaSign(v, pos) and
```

Comment:
I removed this call. We have two cases, either we have a non guarded SSA definition, or we have a guarded one. In case of the latter the Java implementation calls the following predicate too:
```
private Sign guardedSsaSign(Definition v, SsaReadPosition pos) {
  result = ssaDefSign(v) and
  pos.hasReadOfVar(v) and
  hasGuard(v, pos, result)
}
```
which puts two constraints on `result`. `result = ssaDefSign(v)` seems to be too strict, because we might not have a sign for the definition. Consider the following case:
```
if (i < 0)
{
  System.Console.WriteLine(i);
  return;
}
```
We don't have a sign for the `i` in the console writeline through an assignment, but we have one through the guard, which is covered by `hasGuard(v, pos, result)`. So I think we don't need `result = ssaDefSign(v)`. And removing that seems to be equivalent to removing the whole predicate. Maybe there are some performance implications and that's why an extra `pragma[noinline]` is also on that predicate.